GLAA.Repository.Tests/When_interacting_with_generic_entities.cs

```diff
@@ -0,0 +1,314 @@
+using System;
+using System.Linq;
+using GLAA.Common;
+using GLAA.Domain;
+using GLAA.Domain.Models;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NSubstitute;
+
+namespace GLAA.Repository.Tests
+{
+    [TestClass]
+    public class When_interacting_with_generic_entities
+    {
+        private IDateTimeProvider dtp;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            dtp = Substitute.For<IDateTimeProvider>();
+        }
+
+        [TestMethod]
+        public void it_should_retrieve_a_non_deleted_entity_by_id()
+        {
+            const int id = 1;
+            const string expectedAddress = "Line 1";
+
+            var options = new DbContextOptionsBuilder<GLAAContext>()
+                .UseInMemoryDatabase(nameof(it_should_retrieve_a_non_deleted_entity_by_id)).Options;
+
+            using (var context = new GLAAContext(options))
+            {
+                context.Addresses.Add(new Address {Id = id, AddressLine1 = expectedAddress});
+                context.SaveChanges();
+            }
+
+            // For reasons that are unlikely to become clear at the moment, this is best practice
+            using (var context = new GLAAContext(options))
```

Comment:
I think I would agree with treating the EF repo as the class under test so would support it being instantiated. If you're referring to the creation of it's dependencies (GLAAContext) then I can see reasons why. 
I've now looked at the in memory context and like this. 

Would it be possible to somehow have the `InMemoryContext` subbed in place of the GLAA context only during tests? or perhaps have a base setup method which does the options setup so we can re-use the inmemory context with other repo's etc?

Reply:
No. I've seen a few bit of guidance to say that you should isolate the `InMemoryContext` as much as possible so you don't accidentally get one when you actually needed a real database. I'm not sure what you mean about having a base setup method? The advice is to make a new `InMemoryContext` - and therefore repo - for every test (hence the unique names) to ensure there's no cross-contamination of test data.

Reply:
Gotcha, yeh saw the name bit a touch late! 
I was thinking perhaps that the `Setup` method for each test could establish the name of the test which is running (which I think I've done before) (or just use a guid of it's simpler) and create the context as a private property on the class, then just use it. 
Rather than having to create it explicitly inside each test.

Reply:
I see what you mean, yes that makes sense.